tests/handlers/test_room_member.py

```diff
@@ -287,3 +291,119 @@ def test_local_users_joining_on_another_worker_contribute_to_rate_limit(
+        self.bob_token = self.login("bob", "pass")
+
+        # Create a room on this homeserver.
+        self.room_id = self.helper.create_room_as(self.alice, tok=self.alice_token)
+
+    def test_leave_and_forget(self) -> None:
+        """Tests that forget a room is successfully. The test is performed with two users,
+        as forgetting by the last user respectively after all users had left the
+        is a special edge case."""
+        self.helper.join(self.room_id, user=self.bob, tok=self.bob_token)
+
+        # alice is not the last room member that leaves and forgets the room
+        self.helper.leave(self.room_id, user=self.alice, tok=self.alice_token)
+        self.get_success(self.handler.forget(self.alice_ID, self.room_id))
+        self.assertTrue(
+            self.get_success(self.store.did_forget(self.alice, self.room_id))
+        )
+
+        # TODO: server has not forgotten the room
+        # self.assertFalse(
+        #     self.get_success(self.store.is_locally_forgotten_room(self.room_id))
+        # )
+
+    def test_leave_and_forget_last_user(self) -> None:
+        """Tests that forget a room is successfully when the last user has left the room."""
+
+        # alice is the last room member that leaves and forgets the room
+        self.helper.leave(self.room_id, user=self.alice, tok=self.alice_token)
+        self.get_success(self.handler.forget(self.alice_ID, self.room_id))
+        self.assertTrue(
+            self.get_success(self.store.did_forget(self.alice, self.room_id))
+        )
+
+        # TODO: server has forgotten the room
+        # self.assertTrue(
+        #     self.get_success(self.store.is_locally_forgotten_room(self.room_id))
+        # )
+
+    def test_forget_when_not_left(self) -> None:
+        """Tests that a user cannot not forgets a room that has not left."""
+        self.get_failure(self.handler.forget(self.alice_ID, self.room_id), SynapseError)
+
+    def test_rejoin_forgotten_by_server(self) -> None:
+        """Test that a user that has forgotten a room can do a re-join.
+        The room was also forgotten from the local server and only
+        remote users are in the room."""
+        self.get_success(
+            event_injection.inject_member_event(
+                self.hs, self.room_id, "@charlie:elsewhere", "join"
+            )
+        )
+
+        self.helper.leave(self.room_id, user=self.alice, tok=self.alice_token)
+        self.get_success(self.handler.forget(self.alice_ID, self.room_id))
+        self.assertTrue(
+            self.get_success(self.store.did_forget(self.alice, self.room_id))
+        )
+
+        # TODO: server has forgotten the room
+        # self.assertTrue(
+        #     self.get_success(self.store.is_locally_forgotten_room(self.room_id))
+        # )
+
+        self.helper.join(self.room_id, user=self.alice, tok=self.alice_token)
+        self.assertFalse(
+            self.get_success(self.store.did_forget(self.alice, self.room_id))
+        )
```

Comment:
This test is failing
```console
Error: 
Traceback (most recent call last):
  File "/home/runner/work/synapse/synapse/tests/handlers/test_room_member.py", line 376, in test_rejoin_forgotten_by_server
    self.helper.join(self.room_id, user=self.alice, tok=self.alice_token)
  File "/home/runner/work/synapse/synapse/tests/rest/client/utils.py", line 178, in join
    self.change_membership(
  File "/home/runner/work/synapse/synapse/tests/rest/client/utils.py", line 313, in change_membership
    assert channel.code == expect_code, "Expected: %d, got: %d, resp: %r" % (
builtins.AssertionError: Expected: 200, got: 404, resp: b'{"errcode":"M_UNKNOWN","error":"Can\'t join remote room because no servers that are in the room have been provided."}'

tests.handlers.test_room_member.RoomMemberMasterHandlerTestCase.test_rejoin_forgotten_by_server
-------------------------------------------------------------------------------
```

I am not sure if this is a bug in Synapse or a wrong test.
The error occures there:
https://github.com/matrix-org/synapse/blob/0a4efbc1ddc3a58a6d75ad5d4d960b9ed367481e/synapse/handlers/room_member.py#L1668-L1689

Actually, this is not a fault of forgetting rooms, rather of leaving and joining rooms.
Perhaps related to:
- #8796
- #11521

How to deal with it?

Reply:
Synapse doesn't know how to rejoin the user to `self.room_id` because 

- the server has no record of `self.room_id` after it's been forgotten
- the server doesn't know any other servers in that room.

This one might be better handled with a complement test, so that you genuinely do have a remote homeserver; otherwise I think you'd have to do a fair chunk of mocking to do it as a Synapse-only test.

Reply:
> Interesting. The room id is actually one of his own, since his user originally created the room.

The server name in any room ID is not meant to be parsed; it's only there to ensure that two servers don't create a room with the same room ID.
That's why the room ID having been created by this server is actually irrelevant here, I think.

(This surprises people: they assume that `!blahblah:example.com` should be enough to join a room via `example.com` and are confused when Synapse will refuse to if you don't tell it any server names. There are also proposals to remove the server name from room IDs, if I remember, instead using the hash of the create event or something equivalent.)

> A little bit off topic: The server only finds the room again when a local user is invited or with a foreign alias.

This is more or less what I'd expect: in both those cases, the server gets a list of other servers that are in the room.
The server then knows which server it can use to do the `/make_join`, `/send_join` handshake with — see https://spec.matrix.org/v1.3/server-server-api/#joining-rooms if you haven't already to get a picture of how this works.

Joining a remote room requires you to know at least one server that's in the room already, otherwise you have nowhere to join it via.


I agree with @DMRobertson here that a Complement test for this seems like it would be ideal, since it would let you prove this works externally with 2 real servers and you don't have to do a lot of mocking.